cli/flox-rust-sdk/src/providers/build.rs

```diff
@@ -161,3 +161,435 @@ fn read_output_to_channel(
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_second = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_eq!(dir_second, dir_first);
+    }
+
+    #[test]
+    #[ignore] // TODO: Intermittently uses cache.
+    fn build_cache_sandbox_pure_cache_can_be_invalidated() {
+        let package_name = String::from("foo");
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            sandbox = "pure"
+            command = """
+                mkdir $out
+                echo -n {file_content} > $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_first = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        let cache_dir = cache_dir(&env_path, &package_name);
+        assert!(cache_dir.exists());
+        fs::remove_file(cache_dir).unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_second = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_ne!(dir_second, dir_first);
+    }
+
+    #[test]
+    #[ignore] // TODO: Second build fails on `File exists`.
+    fn build_cache_sandbox_off_does_not_use_cache() {
+        let package_name = String::from("foo");
+        let file_name = String::from("bar");
+        let file_content = String::from("some content");
+
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [build.{package_name}]
+            sandbox = "off"
+            command = """
+                mkdir $out
+                echo -n {file_content} > $out/{file_name}
+            """
+        "#};
+
+        let (flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_first = result_dir(&env_path, &package_name).read_link().unwrap();
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        assert_build_file(&env_path, &package_name, &file_name, &file_content);
+        let dir_second = result_dir(&env_path, &package_name).read_link().unwrap();
+        assert_ne!(dir_second, dir_first);
+    }
```

Comment:
This is perfectly common in the real world which is why makefiles use `mkdir -p` pretty much by default. This requirement to leave `$out` behind from previous builds was actually something that I recall Tom insisted upon because it's not uncommon for builds to rely on artifacts found in the target installation directory as they build new versions of the same. I'd call that a broken build, but Tom's right that there are plenty of such examples out there in the wild.

Reply:
Ack on the `mkdir -p`

The resulting issue is that this test assumes that a) without sandbox the build does not use a cache and b) that because it doesn't use a cache the output shall differ.

I think the assumption b) is wrong but then we'll need another way to assert that no cache was used (or that the inplace files a re used as cache effectively).

That would be possible e.g. by 1) writing a random string to a file if the file does not exist then 2) copying that file to $out and 3) asserting that the output of both builds is the same.